rlclientlib/logger/http_transport_client.h

```diff
@@ -126,45 +154,60 @@ pplx::task<web::http::status_code> http_transport_client<TAuthorization>::http_r
   const auto stream = concurrency::streams::bytestream::open_istream(container);
   request.set_body(stream, container_size);
 
-  return _client->request(request).then([this, try_count](pplx::task<http_response> response) {
-    web::http::status_code code = status_codes::InternalError;
-    api_status status;
-
+  // lambda which examines the provided task and either 1) generates a replacement task to retry
+  // the request, or 2) passes the provided task downstream to emit its response
+  auto retry_request_on_failure_lambda = [this, try_count](pplx::task<http_response> response_task) -> pplx::task<http_response>
+  {
+    web::http::status_code response_code = status_codes::InternalError;
+    
     try
     {
-      code = response.get().status_code();
+      response_code = response_task.get().status_code();
     }
     catch (const std::exception& e)
     {
       TRACE_ERROR(_trace, e.what());
     }
 
-    // If the response is not the expected code then it has failed. Retry if possible otherwise report background
-    // error.
-    if (code != status_codes::Created && code != status_codes::NoContent)
+    bool success = (response_code >= status_codes::OK) && (response_code < status_codes::MultipleChoices);
+    if (success)
     {
-      // Stop condition of recurison.
-      if (try_count < _max_retries)
-      {
-        TRACE_ERROR(_trace, "HTTP request failed, retrying...");
+      return response_task;
+    }
 
-        // Yes, recursively send another request inside this one. If a subsequent request returns success we are
-        // good, otherwise the failure will propagate.
-        return send_request(try_count + 1).get();
-      }
-      else
-      {
-        auto msg = u::concat("(expected 201): Found ", code, ", failed after ", try_count, " retries.");
-        api_status::try_update(&status, error_code::http_bad_status_code, msg.c_str());
-        ERROR_CALLBACK(_error_callback, status);
+    // If the response is not success class then it has failed. Retry if possible otherwise report background error.
 
-        return code;
-      }
+    if (try_count >= _max_retries)
+    {
+      // We have exhausted retry attempts, log and return the task describing the failure
+      api_status status;
+      auto msg = u::concat("(expected 201): Found ", response_code, ", failed after ", try_count, " retries.");
+      api_status::try_update(&status, error_code::http_bad_status_code, msg.c_str());
+      ERROR_CALLBACK(_error_callback, status);
+
+      return response_task;
+    }
+
+    if (status_codes::TooManyRequests == response_code)
+    {
+      // if server says it's too busy, delay retry for a while to give it time to recover
+      static int const LINEAR_DELAY_MS_PER_FAILURE = 2000;
+      int delay_ms = (1 + try_count) * LINEAR_DELAY_MS_PER_FAILURE;
+
+      TRACE_ERROR(_trace, u::concat("HTTP request failed with ", response_code, ", retrying in ", delay_ms, "ms..."));      
+
+      concurrency::wait(delay_ms);
```

Comment:
CI is giving an error for `concurrency::wait`. Missing a header?

Reply:
Weird, it built locally for me in visual studio.  Will diagnose what's up.

Reply:
looks like agents async library isn't cross platform but only available for windows

pplx is included via the cpprestsdk is there maybe something from there that could be re-used?

Reply:
would standard library functions work here?
```cpp
#include <thread>
#include <chrono>
std::this_thread::sleep_for(std::chrono::milliseconds(delay_ms));
```

Reply:
@olgavrou, @byronxu99, thank you for this advice.

I dug around a bit and it seems that PPL is not intended to be cross platform, but cpprest does intend to be cross-platform.  cpprest ported to a cross platform pplx the minimal subset which would let them be successful.  In windows pplx just forwards to PPL, so PPL wait() was available during my testing.  In linux I can't see that pplx contains any wait functionality.

@byronxu99 using std sleep will effectively delay my program but it pauses the thread.  It doesn't assign the thread to go contribute to alternate PPLX tasks.  I guess we're not using PPLX deeply inside reinforcement_learning except for the surfaces that touch cpprest, so maybe it's not a big loss.

Let me go research for a while and see if there's any other options besides sleep_for.

Reply:
It appears to me that cpprest and PPLX are currently end-of-life, and don't contain a taskpool aware sleep or yield in their crossplatform library.

I couldn't find any better option than using std::this_thread::sleep to delay the HTTP retry.  The one-second delay I inserted is on the same order as a web request, and since PPLX isn't used much throughout reinforcement_learning we probably won't see much performance difference.  

Instead of sleeping for increasingly long periods of time I reworked the retry logic accept a configured maximum retry timespan, and now all failures retry until either the retry attempts or retry timespan is exhausted.